django/channels/database_layer.py

```diff
@@ -0,0 +1,245 @@
+import base64
+import datetime
+import json
+import random
+import string
+import time
+
+from django.apps.registry import Apps
+from django.db import DEFAULT_DB_ALIAS, connections, models, transaction
+from django.db.utils import OperationalError
+from django.utils import six
+from django.utils.functional import cached_property
+from django.utils.timezone import now
+
+
+class DatabaseChannelLayer(object):
+    """
+    ORM-backed ASGI channel layer.
+
+    For development use only; it will span multiple processes fine,
+    but it's going to be pretty bad at throughput. If you're reading this and
+    running it in production, PLEASE STOP.
+
+    Also uses JSON for serialization, as we don't want to make Django depend
+    on msgpack for the built-in backend. The JSON format uses \uffff as first
+    character to signify a b64 byte string rather than a text string. Ugly, but
+    it's not a valid Unicode character, so it should be safe enough.
+    """
+
+    def __init__(self, db_alias=DEFAULT_DB_ALIAS, expiry=60, group_expiry=86400):
+        self.expiry = expiry
+        self.group_expiry = group_expiry
+        self.db_alias = db_alias
+
+    # ASGI API
+
+    extensions = ["groups", "flush"]
+
+    def send(self, channel, message):
+        # Typecheck
+        assert isinstance(message, dict), "message is not a dict"
+        assert isinstance(channel, six.text_type), "%s is not unicode" % channel
+        # Write message to messages table
+        self.channel_model.objects.create(
+            channel=channel,
+            content=self.serialize(message),
+            expiry=now() + datetime.timedelta(seconds=self.expiry)
+        )
+
+    def receive_many(self, channels, block=False):
+        if not channels:
+            return None, None
+        assert all(isinstance(channel, six.text_type) for channel in channels)
+        # Shuffle channels
+        channels = list(channels)
+        random.shuffle(channels)
+        # Clean out expired messages
+        self._clean_expired()
+        # Get a message from one of our channels
+        while True:
+            try:
+                with transaction.atomic():
+                    message = self.channel_model.objects.filter(channel__in=channels).order_by("id").first()
```

Comment:
Hmmh, if you are using REPEATABLE READ or SERIALIZABLE isolation, then the second session is likely going to fail instead of returning nothing. I believe select_for_update is still the right thing to do.